### ISSUE TITLE
submitOnEnter with nested forms

### ISSUE THREAD

**Issue by basz** [ember-bootstrap/ember-bootstrap#1552]
Output from `ember bootstrap:info`:

```
Npm packages:
ember-bootstrap: ^4.9.0 -> 4.9.0
ember-cli: ~3.24 -> 3.24.0
bootstrap: ^4.5.2 -> 4.5.2
bootstrap-sass: n/a
ember-cli-sass: ^10.0.1 -> 10.0.1
ember-cli-less: n/a
Bower packages:
bootstrap: n/a
bootstrap-sass: n/a
ember-bootstrap configuration:
bootstrapVersion: 4
importBootstrapCSS: false
importBootstrapTheme: false
importBootstrapFont: false
whitelist: undefined
```

I am aware one should create nested forms. However this might prove a useful pattern within ember. I'm using it in a contact components that is actually a dynamicly list of nested specialised forms. Each has there own distinct changeset validations. The containing form might enforce at least one phonenumber, while the individual forms obviously have different validations.

```
form  (@model=contactInfo)
  form  (@model=name)
  form  (@model=phone)
  form  (@model=address)
```

This seems to work fine, except for submit handlers. Especially submitOnReturn doesn't seem to work. As long as the contained form only have 1 field form submission seems to works. The inner form is submitted and validated. However as soon as I have more fields (eg address), the dispatched event (from bs-form) does not reach the intended form. I don't know if this is because of the way the browser handles events (using chrome) or if something else is going on.

But after some digging and experimentation where I bring two modifications to the key handler that seem to be more concise.

- I stop event propagating of the key event so it won't bubble up to any parent element.
- I am exactly targeting the inner form by using the closest('form') selector.

```
 @action
  handleKeyPress(event) {
    let code = event.keyCode || event.which;
    if (code === 13 && this.submitOnEnter) {
      event.stopPropagation();
      let submitEvent = document.createEvent('Event');
      submitEvent.initEvent('submit', true, true);
      event.target.closest('form').dispatchEvent(submitEvent);
    }
  }
```

With these changes my forms work perfectly.

Although my modifications seem logical I worry this change would not be accepted because you guys might feel it is difficult to predict exact behavioural changes across the existing implementations. What do you say?

**Comment by jelhan** [comment]
Nested `<form>` elements are explicitly forbidden by HTML specification:

> § 4.10.3 The form element
>
> **Content model:**
>
> Flow content, but with no form element descendants.
>
> https://html.spec.whatwg.org/multipage/forms.html#the-form-element

I would tend to not add support for it. But I'm wondering if your changes improve the code base even without considering your use case.

Here is a patch view of the changes:

```patch
   handleKeyPress(event) {
     let code = event.keyCode || event.which;
     if (code === 13 && this.submitOnEnter) {
+      event.stopPropagation();
       let submitEvent = document.createEvent('Event');
       submitEvent.initEvent('submit', true, true);
-      event.target.dispatchEvent(submitEvent);
+      event.target.closest('form').dispatchEvent(submitEvent);
     }
   }
```

To questions come to my mind:

1. How do browser handle form submission with `enter` key for forms, which have a submit button? Are both `keypress` and `submit` events fired? Do they both bubble?
2. On which element is the `submit` event dispatched in that case?

I think in all cases the `submitOnEnter` feature should emulate regular browser behavior for forms, which have a submit button as close as possible.

**Comment by basz** [comment]
It's complicated. I found this article outlining the behavior. https://www.tjvantoll.com/2013/01/01/enter-should-submit-forms-stop-messing-with-that/

a single input and no present button[type=submit] *does* trigger a form.submit
multiple inputs and no present button[type=submit] will *not* trigger a form.submit

Which explains my confusion; *no* submit button (or input['type=submit]), vs multiple or single inputs and the expectancy i have for the submitOnEnter property. 

So, while that follows the spec, I would (and I did) - as a user of ember-bootstrap- expect submitOnEnter=true to always submit on enter (when focus is on a (n element of the) form) regardless of the number of element and/or presence of a submit button. Similarly when submitOnEnter=false i would expect the ember-bootstrap to prevent submitting in all cases when enter is pressed.

But I understand to support these cases is kind of complicated. Simple adding an explanation on the behaviour of the submitOnEnter property would do it for me...

So to come back to the above code change. I believe that *if* we want to follow spec we should add a detection for the presence of a submit button and count the number of input elements. Because currently for forms *without* a submit button, a *single* input and submitOnEnter=true two submit events will fire. submitOnEnter=false does now also deviate from the spec if there is a single input.

I guess the becomes is; What kind of experience we want give users of ember-bootstrap. A property like enterOnSubmit can be not set (browser behavior) or set. If set either true or false I honestly expected it to always honour it.
